File: frontend/lib/models/shift.dart
class Shift {
  final String id;
  final String tenantId;
  final String name;
  final String startTime; // formatted as HH:mm:ss
  final String endTime; // formatted as HH:mm:ss
  final int gracePeriodMinutes;
  final int lateRuleMinutes;
  final int earlyRuleMinutes;
  final int overtimeThresholdMinutes;
  final bool isNightShift;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  Shift({
    required this.id,
    required this.tenantId,
    required this.name,
    required this.startTime,
    required this.endTime,
    required this.gracePeriodMinutes,
    required this.lateRuleMinutes,
    required this.earlyRuleMinutes,
    required this.overtimeThresholdMinutes,
    required this.isNightShift,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Shift.fromJson(Map<String, dynamic> json) {
    return Shift(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      name: json['name'] as String,
      startTime: json['start_time'] as String,
      endTime: json['end_time'] as String,
      gracePeriodMinutes: json['grace_period_minutes'] as int? ?? 10,
      lateRuleMinutes: json['late_rule_minutes'] as int? ?? 15,
      earlyRuleMinutes: json['early_rule_minutes'] as int? ?? 15,
      overtimeThresholdMinutes: json['overtime_threshold_minutes'] as int? ?? 30,
      isNightShift: json['is_night_shift'] as bool? ?? false,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tenant_id': tenantId,
      'name': name,
      'start_time': startTime,
      'end_time': endTime,
      'grace_period_minutes': gracePeriodMinutes,
      'late_rule_minutes': lateRuleMinutes,
      'early_rule_minutes': earlyRuleMinutes,
      'overtime_threshold_minutes': overtimeThresholdMinutes,
      'is_night_shift': isNightShift,
      'is_active': isActive,
    };
  }
}

class ShiftSchedule {
  final String id;
  final String tenantId;
  final String employeeId;
  final String shiftId;
  final DateTime effectiveFrom;
  final DateTime? effectiveTo;
  final int? dayOfWeek;
  final DateTime createdAt;
  final String? employeeName;
  final String? shiftName;

  ShiftSchedule({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.shiftId,
    required this.effectiveFrom,
    this.effectiveTo,
    this.dayOfWeek,
    required this.createdAt,
    this.employeeName,
    this.shiftName,
  });

  factory ShiftSchedule.fromJson(Map<String, dynamic> json) {
    return ShiftSchedule(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      shiftId: json['shift_id'] as String,
      effectiveFrom: DateTime.parse(json['effective_from'] as String),
      effectiveTo: json['effective_to'] != null ? DateTime.parse(json['effective_to'] as String) : null,
      dayOfWeek: json['day_of_week'] as int?,
      createdAt: DateTime.parse(json['created_at'] as String),
      employeeName: json['employee_name'] as String?,
      shiftName: json['shift_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tenant_id': tenantId,
      'employee_id': employeeId,
      'shift_id': shiftId,
      'effective_from': effectiveFrom.toIso8601String().substring(0, 10),
      'effective_to': effectiveTo?.toIso8601String().substring(0, 10),
      'day_of_week': dayOfWeek,
    };
  }
}

File: backend/app/schemas/shift.py
"""Shift schemas."""

import uuid
from datetime import time, date, datetime
from typing import Optional
from pydantic import BaseModel, Field, ConfigDict


class ShiftBase(BaseModel):
    name: str = Field(..., max_length=100)
    start_time: time
    end_time: time
    grace_period_minutes: int = Field(default=10, ge=0)
    late_rule_minutes: int = Field(default=15, ge=0)
    early_rule_minutes: int = Field(default=15, ge=0)
    overtime_threshold_minutes: int = Field(default=30, ge=0)
    is_night_shift: bool = False
    is_active: bool = True


class ShiftCreate(ShiftBase):
    pass


class ShiftUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=100)
    start_time: Optional[time] = None
    end_time: Optional[time] = None
    grace_period_minutes: Optional[int] = None
    late_rule_minutes: Optional[int] = None
    early_rule_minutes: Optional[int] = None
    overtime_threshold_minutes: Optional[int] = None
    is_night_shift: Optional[bool] = None
    is_active: Optional[bool] = None


class ShiftResponse(ShiftBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class ShiftScheduleCreate(BaseModel):
    employee_id: uuid.UUID
    shift_id: uuid.UUID
    effective_from: date
    effective_to: Optional[date] = None
    day_of_week: Optional[int] = None


class ShiftScheduleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    employee_id: uuid.UUID
    shift_id: uuid.UUID
    effective_from: date
    effective_to: Optional[date] = None
    day_of_week: Optional[int] = None
    shift_name: Optional[str] = None
    created_at: datetime

    @classmethod
    def model_validate(cls, obj, **kwargs):
        if hasattr(obj, 'shift') and obj.shift:
            data = {
                'id': obj.id,
                'tenant_id': obj.tenant_id,
                'employee_id': obj.employee_id,
                'shift_id': obj.shift_id,
                'effective_from': obj.effective_from,
                'effective_to': obj.effective_to,
                'day_of_week': obj.day_of_week,
                'shift_name': obj.shift.name if obj.shift else None,
                'created_at': obj.created_at,
            }
            return cls(**data)
        return super().model_validate(obj, **kwargs)
